File: lib/models/database/app.database.dart
import 'dart:core';
import 'dart:io';

import 'package:path/path.dart';
import 'package:path_provider/path_provider.dart';
import 'package:sqflite_common_ffi/sqflite_ffi.dart';
import 'package:wallet/helper/logger.dart';
import 'package:wallet/models/account.dao.impl.dart';
import 'package:wallet/models/address.dao.impl.dart';
import 'package:wallet/models/bitcoin.address.dao.impl.dart';
import 'package:wallet/models/contacts.dao.impl.dart';
import 'package:wallet/models/exchangerate.dao.impl.dart';
import 'package:wallet/models/transaction.dao.impl.dart';
import 'package:wallet/models/transaction.info.dao.impl.dart';
import 'package:wallet/models/wallet.dao.impl.dart';

import 'migration.container.dart';
import 'migration.dart';

class AppDatabase {
  static String dbFolder = "databases";
  static String dbName = "proton_wallet_db";
  static String versionKey = "db_version";

  // current version of the database
  int version = 24;

  // future: if the database cached version < resetVersion. rebuild the cache with latest schema. we can clean up migrations.
  int resetVersion = 1;
  bool dbReset = false;

  late Database db;

  late MigrationContainer migrationContainer;
  late AccountDao accountDao;
  late WalletDao walletDao;
  late TransactionDao transactionDao;
  late ContactsDao contactsDao;
  late AddressDao addressDao;
  late BitcoinAddressDao bitcoinAddressDao;
  late TransactionInfoDao transactionInfoDao;
  late ExchangeRateDao exchangeRateDao;

  AppDatabase() {
    migrationContainer = MigrationContainer();
  }

  Future<void> reset() async {
    await dropAllTables();
    await buildDatabase();
  }

  Future<void> dropAllTables() async {
    await walletDao.dropTable();
    await accountDao.dropTable();
    await transactionDao.dropTable();
    await contactsDao.dropTable();
    await addressDao.dropTable();
    await bitcoinAddressDao.dropTable();
    await transactionInfoDao.dropTable();
    await exchangeRateDao.dropTable();
  }

  void initDAO() {
    walletDao = WalletDaoImpl(db);
    accountDao = AccountDaoImpl(db);
    transactionDao = TransactionDaoImpl(db);
    contactsDao = ContactsDaoImpl(db);
    addressDao = AddressDaoImpl(db);
    bitcoinAddressDao = BitcoinAddressDaoImpl(db);
    transactionInfoDao = TransactionInfoDaoImpl(db);
    exchangeRateDao = ExchangeRateDaoImpl(db);
  }

  void buildMigration() {
    final List<Migration> migrations = [
      Migration(1, 2, () async {}),
      Migration(2, 3, () async {
        await transactionDao.migration_0();
      }),
      Migration(3, 4, () async {}),
      Migration(4, 5, () async {}),
      Migration(5, 6, () async {
        await contactsDao.migration_0();
      }),
      Migration(6, 7, () async {
        await addressDao.migration_0();
      }),
      Migration(7, 8, () async {
        await transactionDao.migration_1();
      }),
      Migration(8, 9, () async {
        await transactionDao.migration_2();
      }),
      Migration(9, 10, () async {
        await transactionInfoDao.migration_0();
        await bitcoinAddressDao.migration_0();
      }),
      Migration(10, 11, () async {
        await transactionInfoDao.migration_1();
      }),
      Migration(11, 12, () async {
        await transactionInfoDao.migration_2();
      }),
      Migration(12, 13, () async {}),
      Migration(13, 14, () async {
        await exchangeRateDao.migration_0();
      }),
      Migration(14, 15, () async {
        await transactionDao.migration_3();
      }),
      Migration(15, 16, () async {
        await bitcoinAddressDao.migration_1();
      }),
      Migration(16, 17, () async {
        await transactionDao.migration_4();
      }),
      Migration(17, 18, () async {}),
      Migration(18, 19, () async {}),
      Migration(19, 20, () async {}),
      Migration(20, 21, () async {}),
      Migration(21, 22, () async {
        await walletDao.migration_0();
        await accountDao.migration_0();
      }),
      Migration(22, 23, () async {}),
      Migration(23, 24, () async {
        await accountDao.migration_1();
      }),
      Migration(24, 25, () async {}),
    ];

    migrationContainer.addMigrations(migrations);
  }

  Future<void> init(Database database) async {
    await initDatabase(database);
    buildMigration();
    initDAO();
  }

  static Future<Database> getDatabase() async {
    Database database;
    String dbPath = "";
    if (Platform.isWindows || Platform.isLinux) {
      sqfliteFfiInit();
      final databaseFactory = databaseFactoryFfi;

      final Directory appDocumentsDir =
          await getApplicationDocumentsDirectory();
      dbPath = join(appDocumentsDir.path, dbFolder, dbName);
      database = await databaseFactory.openDatabase(
        dbPath,
      );
    } else {
      final path = await getDatabasesPath();
      dbPath = join(path, dbFolder, dbName);
      database = await openDatabase(
        dbPath,
      );
    }
    logger.d("dbPath: $dbPath");
    return database;
  }

  static Future<Database> getInMemoryDatabase() async {
    Database database;
    if (Platform.isWindows || Platform.isLinux) {
      sqfliteFfiInit();
      final databaseFactory = databaseFactoryFfi;
      database = await databaseFactory.openDatabase(
        inMemoryDatabasePath,
      );
    } else {
      database = await openDatabase(
        ":memory:",
      );
    }
    return database;
  }

  Future<void> initDatabase(Database database) async {
    try {
      if (db.isOpen) {
        logger.i("db is open return");
        return;
      }
    } catch (e) {
      logger.e(e);
    }
    logger.i("set inited Database");
    db = database;
  }

  // TODO(fix): fix me. the oldversion somehow is not correct. we should try to use the version from the database `user_version`.
  // TODO(fix): future we also need to check the db schema is correct or not. then decide to rebuild it or not.
  Future<void> buildDatabase(
      {bool isTesting = false, int oldVersion = 1}) async {
    final List<Migration>? upgradeMigrations =
        migrationContainer.findMigrationPath(oldVersion, version);
    logger.i("Migration appDatabase from Ver.$oldVersion to Ver.$version");
    if (upgradeMigrations != null) {
      for (Migration migration in upgradeMigrations) {
        await migration.migrate();
      }
    } else {
      logger.w("nothing to migrate");
    }

    await checkAndUpdateVersion();
  }

  /// Mark future

  // when db rebuild but table need to resync. this could be in eatch db table
  bool needsResync() {
    return dbReset;
  }

  // not inused for future use
  Future<void> checkAndUpdateVersion() async {
    // Get the current version from the database
    final int currentVersion =
        (await db.rawQuery('PRAGMA user_version')).first.values.first! as int;
    if (currentVersion < version) {
      // If current version is less than the required version
      logger.i(
          "Current version ($currentVersion) is less than required version ($version)");
    }
    await db.execute('PRAGMA user_version = $version');
  }
}

File: rust/src/proton_wallet/db/app_database.rs
use log::info;
use rusqlite::Connection;
use std::sync::Arc;
use tokio::sync::Mutex;

use super::{
    dao::{
        account_dao::AccountDaoImpl, address_dao::AddressDao,
        bitcoin_address_dao::BitcoinAddressDao, contacts_dao::ContactsDao,
        exchange_rate_dao::ExchangeRateDao, proton_user_dao::ProtonUserDao,
        proton_user_key_dao::ProtonUserKeyDaoImpl, transaction_dao::TransactionDao,
        wallet_dao::WalletDaoImpl, wallet_user_settings_dao::WalletUserSettingsDao,
    },
    database::{
        database::BaseDatabase, migration::SimpleMigration, migration_container::MigrationContainer,
    },
    Result,
};
use crate::proton_wallet::db::database::migration::Migration;

#[derive(Debug)]
pub struct AppDatabase {
    pub version: u32,
    pub reset_version: u32,
    pub db_reset: bool,
    pub migration_container: MigrationContainer,
    pub account_dao: AccountDaoImpl,
    pub wallet_dao: WalletDaoImpl,
    pub address_dao: AddressDao,
    pub bitcoin_address_dao: BitcoinAddressDao,
    pub contacts_dao: ContactsDao,
    pub exchange_rate_dao: ExchangeRateDao,
    pub proton_user_dao: ProtonUserDao,
    pub proton_user_key_dao: ProtonUserKeyDaoImpl,
    pub transaction_dao: TransactionDao,
    pub wallet_user_settings_dao: WalletUserSettingsDao,
}

impl Default for AppDatabase {
    fn default() -> Self {
        Self::new("./test_proton_wallet_rust_db.sqlite")
    }
}

impl AppDatabase {
    pub fn new(database_url: &str) -> Self {
        let conn = Arc::new(Mutex::new(Connection::open(database_url).unwrap()));
        AppDatabase {
            version: 3,
            reset_version: 1,
            db_reset: false,
            migration_container: MigrationContainer::new(),
            account_dao: AccountDaoImpl::new(conn.clone()),
            wallet_dao: WalletDaoImpl::new(conn.clone()),
            address_dao: AddressDao::new(conn.clone()),
            bitcoin_address_dao: BitcoinAddressDao::new(conn.clone()),
            contacts_dao: ContactsDao::new(conn.clone()),
            exchange_rate_dao: ExchangeRateDao::new(conn.clone()),
            proton_user_dao: ProtonUserDao::new(conn.clone()),
            proton_user_key_dao: ProtonUserKeyDaoImpl::new(conn.clone()),
            transaction_dao: TransactionDao::new(conn.clone()),
            wallet_user_settings_dao: WalletUserSettingsDao::new(conn.clone()),
        }
    }

    pub async fn reset(&self) -> Result<()> {
        self.drop_all_tables().await?;
        self.build_database(1).await
    }

    pub async fn drop_all_tables(&self) -> Result<()> {
        let _ = self.account_dao.database.drop_table().await;
        Ok(())
    }

    pub fn build_migration(&mut self) {
        let migrations = vec![
            SimpleMigration::new(1, 2, {
                let wallet_database = self.wallet_dao.database.clone();
                let account_database = self.account_dao.database.clone();
                let address_database = self.address_dao.database.clone();
                let bitcoin_address_database = self.bitcoin_address_dao.database.clone();
                let contacts_database = self.contacts_dao.database.clone();
                let exchange_rate_database = self.exchange_rate_dao.database.clone();
                let proton_user_database = self.proton_user_dao.database.clone();
                let proton_user_key_database = self.proton_user_key_dao.database.clone();
                let transaction_database = self.transaction_dao.database.clone();
                let wallet_user_settings_database = self.wallet_user_settings_dao.database.clone();
                move || {
                    let wallet_database = wallet_database.clone();
                    let account_database = account_database.clone();
                    let address_database = address_database.clone();
                    let bitcoin_address_database = bitcoin_address_database.clone();
                    let contacts_database = contacts_database.clone();
                    let exchange_rate_database = exchange_rate_database.clone();
                    let proton_user_database = proton_user_database.clone();
                    let proton_user_key_database = proton_user_key_database.clone();
                    let transaction_database = transaction_database.clone();
                    let wallet_user_settings_database = wallet_user_settings_database.clone();
                    async move {
                        let _ = wallet_database.migration_0().await;
                        let _ = account_database.migration_0().await;
                        let _ = address_database.migration_0().await;
                        let _ = bitcoin_address_database.migration_0().await;
                        let _ = contacts_database.migration_0().await;
                        let _ = exchange_rate_database.migration_0().await;
                        let _ = proton_user_database.migration_0().await;
                        let _ = proton_user_key_database.migration_0().await;
                        let _ = transaction_database.migration_0().await;
                        let _ = wallet_user_settings_database.migration_0().await;
                    }
                }
            }),
            SimpleMigration::new(2, 3, {
                let account_database = self.account_dao.database.clone();
                move || {
                    let account_database = account_database.clone();
                    async move {
                        let _ = account_database.migration_1().await;
                    }
                }
            }),
        ];
        self.migration_container.add_migrations(migrations);
    }

    pub async fn init(&mut self) -> Result<()> {
        self.build_migration();
        Ok(())
    }

    pub async fn build_database(&self, old_version: u32) -> Result<()> {
        let upgrade_migrations = self
            .migration_container
            .find_migration_path(old_version, self.version);

        info!(
            "Migration appDatabase from Ver.{} to Ver.{}",
            old_version, self.version
        );

        if let Some(migrations) = upgrade_migrations {
            for migration in migrations {
                migration.migrate().await;
            }
        } else {
            info!("nothing to migrate");
        }

        self.check_and_update_version().await
    }

    pub async fn check_and_update_version(&self) -> Result<()> {
        // TODO:: use rust local storage to save version
        Ok(())
    }

    pub fn needs_resync(&self) -> bool {
        self.db_reset
    }
}
